### ISSUE TITLE
cross platform support (e.g. with CMake)?

### ISSUE THREAD

**Issue by pieper** [madler/pigz#62]
We'd be interested in using pigz as a library in [our application](slicer.org) and would want to support Windows, Linux, and Mac with a uniform build system that supports native compilers (e.g. Visual Studio, not mingw).  CMake has been our go-to solution for that so it would be great if pigz had CMake support.

Has anyone looked into CMake for pigz before?

**Comment by neurolabusc** [comment]
@pieper can you try the branch below which adapts @ningfei's Cmake super build scripts. These scripts have been used for years with dcm2niix. One clever feature is you can specify the ZLIB_IMPLEMENTATION to be `System` or `Cloudflare` (the default). The [Cloudflare zlib](https://github.com/cloudflare/zlib) accelerates pigz 40%, though it requires an Intel CPU built since 2009. 

You can try this out with the following command:

```
git clone https://github.com/neurolabusc/pigz.git
cd pigz
mkdir build && cd build
```

I would be happy if @madler would consider including this Cmake script in the main pigz repository (though he might choose whether the System zlib should be the default).

Since the Cmake uses the glibc of your system, it would be great to test if these address [issue 68](https://github.com/madler/pigz/issues/68). None of my computers have ever exhibited that issue, it would be great to know that it has been addressed.

**Comment by neurolabusc** [comment]
@pieper the current CMake scripts support Unix (MacOS and Linux), but not Windows. There is a [fork of pigz](https://github.com/kjk/pigz) that supports Windows, albeit v2.3.1 versus v2.4. @ningfei's dcm2niix scripts do support Windows, so maybe he can can extend this to pigz. It would be nice to have Windows support be part of the main trunk and even better if users can leverage the Cloudflare zlib.

**Comment by pieper** [comment]
See also the discussion here: https://discourse.vtk.org/t/cloudflare-zlib/2330/7

**Comment by dzenanz** [comment]
`pigz` is a program and not a library, therefore it cannot really replace `zlib`.

**Comment by neurolabusc** [comment]
@dzenanz, you are correct that `pigz` is a program and not a library. However, on Linux you can use it piped mode. After calling `popen` you can write to pigz using the same routines that you would to write to an uncompressed file, closing it with `pclose`.   Please take a look at my modification of [znzlib](https://github.com/neurolabusc/znzlib) where a few lines of code allows one to select pigz output using the same functions designed to write uncompressed data. That project demonstrates how the a copy of pigz compiled with CloudFlare zlib provides optimized parallel compression with very little changes to the code and no changes to the make script. In my example, I simply detect an environment variable to see if the user wants to write the data to pigz.

**Comment by dzenanz** [comment]
[About 50%](https://hostingtribunal.com/blog/operating-systems-market-share/#content-4) of programmers use Windows, and that percentage is much greater for ordinary users. If you don't have to support Windows (for whatever reason), PIGZ is fine. If you do (as all multi-platform libraries aim to do), PIGZ is out of the question.

**Comment by neurolabusc** [comment]
@dzenanz point taken. I evaluated znzlib since it is used by several of the most popular tools in my field (AFNI, FSL, FreeSurfer) and those tools only target Unix (though many Windows users run them via Windows Subsystem for Linux). I do not people have ported pigz to Windows, though I do not know if they have added modern CMake support or if pipes work with those. Using a pipe does have performance benefits. @ningfei is a Cmake guru who developed the [dcm2niix](https://github.com/rordenlab/dcm2niix) scripts to work on Windows, Linux and MacOS as well as the Unix pigz CMake from my repository. Your comments might give him the incentive to look at making a CMake script for pigz that includes all the operating systems.

**Comment by ningfei** [comment]
Not really a CMake guru :D As I searched, currently there are only [binaries](http://binaries.przemoc.net/) available for the latest pigz for Windows. The open source [port](https://github.com/kjk/pigz) might be not up-to-date. Will have a look when I have time.

**Comment by neurolabusc** [comment]
@ningfei thanks for looking into this. As you note, forks of pigz that support the Microsoft compiler are based on [pigz 2.3.1](https://github.com/kjk/pigz), which predates piped storage and the notes go up to Visual Studio 2012 (and remake instead of CMake). There are Windows compatible executables for [pigz 2.4](http://binaries.przemoc.net) built using gcc on MSYS2. I know the recent versions of Microsoft compilers have improved their compatibility with GCC conventions. Popular tools in my field (FSL, AFNI) only support Unix (though they run fine on Windows Subsystem for Linux), and the vtk thread linked above suggests the developers of that tool prefer libraries over stand alone applications. 

@madler can you tell us how to proceed. To summarize @pieper asked for an `enhancement` for pigz to support CMake for deploying to MacOS, Linux and Windows. @ningfei and I provide a [CMake](https://github.com/neurolabusc/pigz) script that supports MacOS and Linux (and optionally optimized versions of zlib). However, this script does not support Windows. The options are:
 1. If you request, I can generate a pull request to add the CMake script to your repository.
 2. You could relabel this issue as an `enhancement` and keep it open until some developer steps up to extend the CMake to support Windows.
 3. You could close the issue, as it is an enhancement and not a bug, and outside of the remit of your project.

**Comment by madler** [comment]
Sure, you can generate a pull request for a CMake script, and I'll include it. However it sounds like your script does not resolve the original request for Windows support.

**Comment by neurolabusc** [comment]
You are correct, the CMake script currently only supports MacOS and Linux. It has the general benefits of CMake (detecting capabilities, installing missing components) and liabilities (requiring CMake), plus the added benefit of simplifying building with optimized zlib. So it would be a simple alternative to the current make file. 

Before I make a pull request, I make a call out to Windows developers to see if anyone wants to extend the CMake script to Windows. @kjk do you have any interest in this?

**Comment by neurolabusc** [comment]
@madler I need some guidance on how to proceed with the pull request. There are three options:

 1. The existing [CMake](https://github.com/neurolabusc/pigz) makes minimal changes to pigz, targets Unix and MacOS. Allows the user to select CloudFlare-zlib, System or zlib-ng. Since the core code is virtually untouched, this has low risk of unintended consequences.
 2. The `windows` branch includes the minimal changes to compile to Windows (while retaining compatibility with MacOS and Linux). At the moment, this can be built from the command line, but I think @ningfei could get a CMake working with Windows pretty easily. Since the core code is virtually untouched, this also has a low risk of unintended consequences. However, it does have a clear known limitation: it requires that filenames use Latin characters, and will fail with non-Latin values, e.g. [pigz Описание.pdf](https://gist.github.com/przemoc/bd6c342ff7f31a0c4ec2) returns `pigz.exe: skipping: ????????.pdf does not exist`.

 3. I am willing to create a more thorough modification that will handle Windows files. But this course is only worth pursuing if you are willing to thoroughly vet the changes and would consider incorporating the changes in the main release. Given the large Windows community, I do think bugs could be quickly identified, but the code will have a lot of `ifdefs` that will make it harder to maintain. Specifically, the `char *` type will need to be `wchar_t * `. I think this is the simplest robust solution, but it will create widespread changes throughout the code. Below is a minimal example to demonstrate the changes. 

```
// arg.c
// compile with: 
// cl /EHsc /TP arg.c
// arg  Описание.pdf

#include <io.h> // _setmode()
#include <fcntl.h> //_O_U16TEXT
#include <stdio.h>
                        
// For local functions and globals.
#define local static

#if defined(_MSC_VER)
local void process(wchar_t* arg) {
 wprintf(L"processing '%wS'\n", arg);
#else
local void process(char *path) {
 printf("processing '%s'\n", arg);
#endif
}

#if defined(_MSC_VER)
int wmain(int argc, wchar_t* argv[]) {
#else
int main(int argc, char **argv) {
#endif
    _setmode(_fileno(stdout), _O_U16TEXT);
    for (int i = 1; i < argc; ++i) {
    	process(argv[i]);
    }
 }
```

I am happy to follow any path you recommend, would just like to know your preference. I appreciate all your work to the community, and respect that ultimately you must maintain this code.

**Comment by kjk** [comment]
@neurolabusc to minimize use of #ifdef and amount of changes I would recommend using utf8 on windows

Do WCHAR -> utf8 at the beginning of wmain.

Do utf8 => WCHAR conversions in leaf functions like "file open".  That way most of the code is unchanged.

**Comment by neurolabusc** [comment]
@kjk would you be interested in trying to implement this? I was worried there would be a lot of changes, including the `--name` storage and in `try.c`. Maybe I just have the wrong approach. You could make a fork of my branch, will has the minimal Windows support without handling non-Latin characters. Here are the commands to download my fork and compile on Windows (assuming you are running the "x64 Native Tools Command Prompt":

```
 git clone --branch windows https://github.com/neurolabusc/pigz
 cd pigz\src\win
 msvc.bat
```

I have very little experience with MSVC, so having someone with more experience would be greatly appreciated.

**Comment by neurolabusc** [comment]
@kjk, @dzenanz and @pieper I would be grateful if you could try out my fork's latest [release](https://github.com/neurolabusc/pigz/releases/tag/v2.4.win). I have implemented @kjk's suggestion of converting WCHAR -> utf8 at the beginning of wmain. This results with minimal changes and in my testing handles both Latin and non-Latin filenames. Since utf8 is used internally, I believe the `--name` works as intended. I am not an experienced Windows programmer, and would appreciate any suggestions as well as testing.

Assuming everyone is happy with the Windows compilation, the final step is to implement CMake for Windows. I have updated @ningfei's CMake script to compile with `System`, `Cloudflare`, `Intel `or `ng ` variants of zlib. I have also created a simple script to compile, test and [benchmark](https://github.com/neurolabusc/pigz-bench) these variants. The file `msvc.bat` shows how simple it is to compile pigz on Windows, but my experience with CMake is very limited. I realize you are busy, but if @ningfei or @pieper can help update CMake to handle Windows, I think this will result in a compelling pull request.

**Comment by neurolabusc** [comment]
@jcfr would you be interested in helping @pieper out on this. I notice you do a lot of the CMake commits for your joint tools. The new pigz [CMake scripts](https://github.com/neurolabusc/pigz/) work on Linux and MacOS, and a tiny [command line script](https://github.com/neurolabusc/pigz/blob/master/src/win/msvc.bat) compiles this for Windows. However, I am very unfamiliar with CMake, and it is unclear to me how to link the pthread library for Windows. Failing that, can you suggest a Windows CMake guru who can help out. I notice that other tools like Blender use CMake to link the Pthread library on Windows.

**Comment by pieper** [comment]
Hi @neurolabusc , sorry to leave you hanging on this but it's not really in my critical path and I know how much time it can take to iron out all the details when you mix windows, cmake, and legacy code.  It's not that it cannot be done, thing just tend to pop up over time so long term maintenance may always be an issue (for example we keep a [fork of libarchive](https://github.com/Slicer/libarchive) in Slicer that ends up needing debugging from time to time on various platforms.

**Comment by neurolabusc** [comment]
@pieper it should be fully functional now. I no longer need any help. I have submitted a [Pull Request](https://github.com/madler/pigz/pull/77) so just waiting for @madler to review.

**Comment by sergeevabc** [comment]
Could you be so kind to generate the latest ```.exe``` for the rest of us who are mere Windows users w/o compiler?

**Comment by rafadess** [comment]
I'm really looking forward to see the PR implemented to the source code. We use that [windows implementation of pigz](https://github.com/kjk/pigz) but it is kind of outdated. So would be awesome if we could build for windows from the original repo. I'm subscribing this topic and the PR to receive any news.
Thank you for this implementation.

**Comment by madler** [comment]
If someone would like to make an .exe, I can include it in the distribution. I don't have a Windows machine.

**Comment by neurolabusc** [comment]
@madler  you can get a Windows executable from here:
  https://github.com/neurolabusc/pigz/releases
This was built using [pull request 77](https://github.com/madler/pigz/pull/77). This pull request has been waiting one year for review. I realize free projects rely on people's free time, but please tell me if there is any way I can expedite the review of that PR. It provides a CMake system, provides Windows threading, and supports Windows UTF16 filenames. It does not change any of the internal logic, so I think the chance of unintended consequences is low.

**Comment by ningfei** [comment]
Not fully tested but two versions here: one built using MingW GCC 10.2, one built using Visual Studio 2019 Version 16.8. Both based on the latest code of pigz 2.6 and PR #77 (I merged them locally), using the CloudFlare zlib with intrinsics enabled. I've actually been trying to setup a CI pipeline so we can easily get the executables, but don't have enough time to finish it...
[pigz.zip](https://github.com/madler/pigz/files/6032069/pigz.zip)

**Comment by dieperdev** [comment]
@neurolabusc Do you have an executable with large file support? I'm getting this error when I try to compress a 40gb file:
![image](https://github.com/madler/pigz/assets/140034404/002bd884-64cc-4728-a502-f8257cb74d7d)

**Comment by neurolabusc** [comment]
@dieperdev I submitted my pull request to support Windows appropriately four years ago, and it has yet to be reviewed. My sense is that supporting this platform is not a priority for the lead developer. I really only want to support this if it can be made part of the core code base to support all users and be maintained. Therefore, I am focusing my development efforts elsewhere. Happy to renew my work if @madler or another maintainer wants to merge my contributions or suggest how my work could be improved to allow a merge.

**Comment by sergeevabc** [comment]
> @ningfei
> Not fully tested but two versions here: [pigz.zip](https://github.com/madler/pigz/files/6032069/pigz.zip)

Both of them crash on my end (Core2Duo with SSE3).
Not sure, but maybe due to the need for AVX instructions.

> @neurolabusc
> you can get a Windows executable from here: https://github.com/neurolabusc/pigz/releases

pigz_Windows.zip works,
pigz_Windows_CloudFlareZlib.zip crashes.

**Comment by DanielRuf** [comment]
@neurolabusc maybe the repo needs an update, since pigz 2.8 is available now.

Also I think it would be helpful to have a GitHub Actions setup that compiles the binary (for tagged commits). With this we don't need a full local development environment just for compiling the new binaries.

**Comment by DanielRuf** [comment]
Personally I would like to see a setup like GNU uses which is quite simple (using bootstrap, config and make).

https://github.com/DanielRuf/gzip-windows-build/blob/main/.github/workflows/main.yml

**Comment by neurolabusc** [comment]
@DanielRuf it has been 53 months since I submitted [PR#77](https://github.com/madler/pigz/pull/77). I am reluctant to invest time if there is no chance of my efforts being merged - I did target the current pigz when I made my PR. I am happy to put effort into this if there is a chance that the subsequent PR has a chance of being part of the main project. @madler can you provide any guidance on how to proceed. I think many would benefit if pigz could support non-Latin characters for the Windows operating system.

**Comment by leanhdung1994** [comment]
I hope pigz can come to Windows. I hope @neurolabusc's request https://github.com/madler/pigz/pull/77 can be reviewed and merged...

**Comment by neurolabusc** [comment]
@madler I can rebase [PR77](https://github.com/madler/pigz/pull/77) on the current master. Since my submission in 69 months ago it has fallen 59 commits behind master. I would be happy to update it, if it has a reasonable chance of being merged. We could also add appveyor as [seen here](https://github.com/ningfei/pigz) that could compile for various targets (including Windows) and could run regression tests (including files with non-Latin characters). Just tell me how to proceed.

**Comment by leanhdung1994** [comment]
@neurolabusc you have my highest respect for your persistence and dedicated support.

Given the current siguation, do you have any suggestion for parallel compression into gzip on Windows? Even the popular 7-ZIP does not support this feature: https://github.com/mcmilk/7-Zip-zstd/issues/444.

**Comment by neurolabusc** [comment]
@leanhdung1994 sorry, I can not suggest a multi-threaded alternative for GZip.

If you want compression, but do not specifically require the GZip format, consider zstd. 

If you need GZip, the single-threaded performance of [libdeflate](https://github.com/ebiggers/libdeflate) for both compression and decompression is really impressive and may be sufficient for your needs. Beyond speed, it also allows you to select superior compression size (up to `12`) that is useful for creating files for web pages that will only be compressed once but transferred and decompressed many times. 

Below is the performance of different tools on my 14-core Apple M4 Pro using my [Python benchmark](https://github.com/neurolabusc/pigz-bench/tree/python). The Pareto frontier is to be at the bottom (small files) and toward the right (fast compression).

<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/40795c39-af14-465c-9c06-d7c93ec3baac" />

**Comment by leanhdung1994** [comment]
@neurolabusc 
I am process the HTML dumps of Wikipedia where all upstream files are gzip, so gzip is my only option.

I am aware of zstd built-in support for parallel processing. I wrote an email to Wikimedia to suggest them to replace gzip with zstd.

**Comment by neurolabusc** [comment]
@leanhdung1994 I do not have a Windows machine locally, but can you build my branch. Not sure if you use cmake or the msvc.bat file. The cmake script would look something like this:

```
git clone https://github.com/neurolabusc/pigz
cd pigz
mkdir .\build
cd .\build
cmake -Wno-dev -G "Visual Studio 16 2019 Win64" ..\
cmake --build . --config Release
```

**Comment by leanhdung1994** [comment]
@neurolabusc I have found [`pgzip`](https://github.com/pgzip/pgzip) which works great on Python 3.13.5 in Windows 11:

````python
from pathlib import Path
import shutil, pgzip

outDir  = r"E:\Personal Projects\tmp"
outDir  = Path(outDir)
inTar   = outDir / "chunk_0.tar"
gzipDir = outDir / "chunk_0.tar.gz"

with open(inTar, 'rb') as f, pgzip.open(gzipDir, "wb", thread=10, blocksize=2*10**8) as g:
        shutil.copyfileobj(f, g)
````

and

````python
from pathlib import Path
import tarfile, pgzip

# List of NDJSON files you want to compress
fileToCompress = [r"E:\Personal Projects\tmp\chunk_0.ndjson",
                  r"E:\Personal Projects\tmp\chunk_1.ndjson"]

fileToCompress = [Path(p) for p in fileToCompress]

# Output archive name
outputArchive = r"E:\Personal Projects\tmp\dataset.tar.gz"
outputArchive = Path(outputArchive)

with pgzip.open(outputArchive, "wb", thread=10, blocksize=64*1024) as f, tarfile.open(fileobj=f, mode="w:") as g:
    for file in fileToCompress:
        # Use only the filename (no directories) inside the archive
        g.add(file, arcname=file.name)
````

On the other hand, there is a [problem](https://github.com/pgzip/pgzip/issues/43) of installation on Python 3.14.2.

**Comment by neurolabusc** [comment]
@leanhdung1994 that is a nice find. 

Clear strengths of pgzip: 
 - Unlike pigz, it is able to handle [non-Latin characters](https://gist.github.com/przemoc/bd6c342ff7f31a0c4ec2) in Windows (which uses UTF16, not UTF8). Therefore, it is an alternative to [our PR](https://github.com/madler/pigz/pull/77).
 - It creates a gzip file that it can decompress in parallel, but is still compatible with other serial gzip decompression. Though it is only able to parallel decompress files it creates.

Potential weakness:
 - On my 14-core MacOS computer using Python 3.13.2, this Python script did not perform as well as suggested on the [project web page](https://github.com/pgzip/pgzip?tab=readme-ov-file#compression-performance). I tried using default (automatic) and explicit numbers of threads (which gave the same results). I tried the default (100mb) and the [tested](https://github.com/pgzip/pgzip) 200mb block size (which gave the same results). Performance using the 278mb Silesia Corpus using the defaults is shown below. 

<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/77c5160c-99eb-4394-9296-2c8529504b02" />

In my testing, peak performance was found by setting the block size to input_size/threads (see below). If you do use this Python compressor, you will probably want to customize the block size based on the file size, number of threads and your available RAM. In addition to non-Latin characters, @ningfei's contribution to [our PR](https://github.com/madler/pigz/pull/77) also allows using the zlib-ng library, and as shown below this unlocks additional performance.
<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/c120672f-d19d-4a6b-9590-afb81e7320d2" />

**Comment by leanhdung1994** [comment]
@neurolabusc In my experiments (compressing a .tar file of 2048MB), I get

<img width="1255" height="922" alt="Image" src="https://github.com/user-attachments/assets/1b1e1770-defa-434b-916d-63013c91ceed" />

and

<img width="1248" height="928" alt="Image" src="https://github.com/user-attachments/assets/511ded51-6145-4da6-a680-e508bfbac4b8" />

On the other hand, I have found the GO package [`pgzip`](https://github.com/klauspost/pgzip), whose Github repo has 1.200 stars. The fact that this package for GO makes it less compelling for my all-Python pipeline. Below is the Python script I used for the experiments:

````python
from pathlib import Path
import shutil, time, pgzip
import matplotlib.pyplot as plt

outDir  = r"E:\Personal Projects\tmp"
outDir  = Path(outDir)
tarDir  = outDir / "chunk_0.tar"
gzipDir = outDir / "chunk_0.tar.gz"

cores   = []
times   = []
ratios  = []
speeds  = []

for nCore in range(1, 10):
    if gzipDir.exists():
        gzipDir.unlink()

    start = time.perf_counter()

    with open(tarDir, 'rb') as f, pgzip.open(gzipDir, "wb", thread=nCore, blocksize=2*10**8) as g:
            shutil.copyfileobj(f, g)

    end     = time.perf_counter()
    runtime = end - start
    ratio   = gzipDir.stat().st_size / tarDir.stat().st_size
    speed   = (tarDir.stat().st_size / 1024 / 1024) / runtime
    cores.append(nCore)
    times.append(runtime)
    ratios.append(ratio)
    speeds.append(speed)

fig, ax1 = plt.subplots()

ax1.set_xlabel("Number of threads")
ax1.set_ylabel("Speed (MB/s)")
ax1.plot(cores, speeds, marker="o")

plt.title("pgzip Compression: Threads vs Speed")
plt.tight_layout()
plt.show()

fig, ax1 = plt.subplots()

ax1.set_xlabel("Number of threads")
ax1.set_ylabel("Running time (seconds)")
ax1.plot(cores, times, marker="o")

plt.title("pgzip Compression: Threads vs Time")
plt.tight_layout()
plt.show()
````

**Comment by leanhdung1994** [comment]
@neurolabusc I have tested the GO package https://github.com/klauspost/pgzip with the same file as in https://github.com/madler/pigz/issues/62#issuecomment-3694952423. Its speed is impressive:

<img width="1517" height="1025" alt="Image" src="https://github.com/user-attachments/assets/1dbd1316-be17-435a-aa30-f275ee2c2648" />

and

<img width="1668" height="1119" alt="Image" src="https://github.com/user-attachments/assets/d61d2145-3d10-4c5b-967c-4377f12549cf" />

Below is the code I used (from ChatGPT and Grok):

````go
package main

import (
	"fmt"
	"io"
	"os"
	"time"

	"github.com/klauspost/pgzip"
	"gonum.org/v1/plot"
	"gonum.org/v1/plot/plotter"
	"gonum.org/v1/plot/vg"
	"gonum.org/v1/plot/vg/draw"
)

func main() {
	tarPath := `E:\Personal Projects\tmp\chunk_0.tar`
	gzipPath := `E:\Personal Projects\tmp\chunk_0.tar.gz`

	var threads []float64
	var speeds []float64
	var times []float64

	for nThreads := 1; nThreads <= 9; nThreads++ {
		// Remove previous gzip file
		os.Remove(gzipPath)

		start := time.Now()

		tarFile, err := os.Open(tarPath)
		if err != nil {
			panic(err)
		}

		outFile, err := os.Create(gzipPath)
		if err != nil {
			panic(err)
		}

		gz, err := pgzip.NewWriterLevel(outFile, pgzip.DefaultCompression)
		if err != nil {
			panic(err)
		}

		gz.SetConcurrency(1<<20, nThreads)

		if _, err := io.Copy(gz, tarFile); err != nil {
			panic(err)
		}

		gz.Close()
		tarFile.Close()
		outFile.Close()

		elapsed := time.Since(start).Seconds()

		tarInfo, _ := os.Stat(tarPath)
		gzInfo, _ := os.Stat(gzipPath)

		sizeMB := float64(tarInfo.Size()) / 1024.0 / 1024.0
		speed := sizeMB / elapsed
		ratio := float64(gzInfo.Size()) / float64(tarInfo.Size())

		fmt.Printf("%d, Time: %.3f s, Speed: %.2f MB/s, Ratio: %.3f\n", nThreads, elapsed, speed, ratio)

		threads = append(threads, float64(nThreads))
		speeds = append(speeds, speed)
		times = append(times, elapsed)
	}

	// Plot Threads vs Speed
	plotLine(threads, speeds, "Threads", "Speed (MB/s)", "pgzip Compression: Threads vs Speed", "speed_plot.png")

	// Plot Threads vs Time
	plotLine(threads, times, "Threads", "Time (seconds)", "pgzip Compression: Threads vs Time", "time_plot.png")
}

// plotLine creates a PNG line plot from x and y slices
func plotLine(x, y []float64, xlabel, ylabel, title, filename string) {
	p := plot.New()
	p.Title.Text = title
	p.X.Label.Text = xlabel
	p.Y.Label.Text = ylabel

	pts := make(plotter.XYs, len(x))
	for i := range x {
		pts[i].X = x[i]
		pts[i].Y = y[i]
	}

	line, err := plotter.NewLine(pts)
	if err != nil {
		panic(err)
	}
	line.Color = plotter.DefaultLineStyle.Color
	p.Add(line)

	// Add points
	scatter, err := plotter.NewScatter(pts)
	if err != nil {
		panic(err)
	}
	scatter.GlyphStyle = draw.GlyphStyle{
		Shape:  draw.CircleGlyph{},
		Color:  scatter.Color, // optional
		Radius: vg.Points(3),  // size of the marker
	}
	p.Add(scatter)

	// Save PNG
	if err := p.Save(6*vg.Inch, 4*vg.Inch, filename); err != nil {
		panic(err)
	}
	fmt.Println("Plot saved to", filename)
}

````

**Comment by leanhdung1994** [comment]
@neurolabusc  my [fork](https://github.com/leanhdung1994/pgzip) of [pgzip](https://github.com/pgzip/pgzip) improves compression performance significantly. I managed to reduce compression time from 27s to 14s on an 8.77GB file (same settings: ` thread=5, blocksize=10**7, compresslevel=6`).
